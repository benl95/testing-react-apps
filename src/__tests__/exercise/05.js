// mocking HTTP requests
// http://localhost:3000/login-submission

import * as React from 'react'
import {render, screen, waitForElementToBeRemoved} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {build, fake} from '@jackfranklin/test-data-bot'
import {setupServer} from 'msw/node'
import {handlers} from '../../test/server-handlers'
import Login from '../../components/login-submission'

function getInputFieldsFromForm(...labels) {
  return labels.map((label, _) => screen.getByLabelText(label))
}

const buildLoginForm = build({
  fields: {
    username: fake(f => f.internet.userName()),
    password: fake(f => f.internet.password()),
  },
})

const server = setupServer(...handlers)

beforeAll(() => server.listen())
afterAll(() => server.close())

test(`logging in displays the user's username`, async () => {
  render(<Login />)
  const {username, password} = buildLoginForm()
  const [usernameInputField, passwordInputField] = getInputFieldsFromForm(
    /username/i,
    /password/i,
  )

  await userEvent.type(usernameInputField, username)
  await userEvent.type(passwordInputField, password)

  await userEvent.click(screen.getByRole('button', {name: /submit/i}))

  await waitForElementToBeRemoved(() => screen.getByLabelText(/loading/i))

  expect(screen.getByText(username)).toBeInTheDocument()
})

test('omitting the username results in an error', async () => {
  render(<Login />)
  const {password} = buildLoginForm()
  const [passwordInputField] = getInputFieldsFromForm(/password/i)

  // don't type in the username
  await userEvent.type(passwordInputField, password)

  await userEvent.click(screen.getByRole('button', {name: /submit/i}))

  await waitForElementToBeRemoved(() => screen.getByLabelText(/loading/i))

  expect(screen.getByRole('alert')).toHaveTextContent(/username required/i)
})
