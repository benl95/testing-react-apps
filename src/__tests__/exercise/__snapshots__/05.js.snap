// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`omitting the username results in an error 1`] = `
<div>
  username required
</div>
`;
